File: week_8_and_9/3_JavaScript/grocery_list/my_solution.js
// U3.W8-9: 


// I worked on this challenge [by myself, with: ].

// 2. Pseudocode

// # def class GroceryList
// # initialize instance variable @list_hash
// # define a method for adding items to list
// # will require knowing item and quantity
// # will be added to @list_hash
// # define a method for removing items 
// # will require item name
// # delete the item from @list_hash
// # define a method for adjusting the quantity
// # will name and quantity
// # change the value of given key to the new quantity
// # define a method that displays the list(item and quantity)
 

// 3. Initial Solution

var groceryList= {};
groceryList['list'] = {};

groceryList.add_item = function(item, qty) {
	groceryList['list'][item] = qty;
}
groceryList.remove_item = function(item) {
	delete groceryList['list'][item]
}
groceryList.change_qty = function(item, qty) {
	groceryList['list'][item] = qty; 
}

 
// 4. Refactored Solution

var groceryList = {
	"list": {},
	add_item: function(item, qty) {
		groceryList['list'][item] = qty;
	},
	remove_item: function(item) {
		delete groceryList['list'][item]
	},
	change_qty: function(item, qty) {
		groceryList['list'][item] = qty; 
	}
}




// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

function assert(test, message, number) {
  if (!test) {
    console.log(number + "false");
    throw ("Error" + message);  
  } 
    console.log(number + "true");
    return true;   
 }
 
assert(
   (groceryList.add_item instanceof Function),
   "groveryList.add_item should be defined as a function", 
   "1. "
);

assert(
   (groceryList.remove_item instanceof Function),
   "groveryList.remove_item should be defined as a function", 
   "2. "
);

// assert(
//    (groceryList.display_list instanceof Function),
//    "groveryList.display_list should be defined as a function", 
//    "3. "
// );

groceryList.add_item('avocado', 2);
groceryList.add_item('chiccken', 1);
groceryList.add_item('bread', 5);



assert(
   (Object.keys(groceryList.list).length == 3),
   "groceryList should have three items in its list", 
   "4. "
);

assert(
   (groceryList.list.bread == 5),
   "bread in the groceryList should equal to 5", 
   "5. "
);

groceryList.remove_item('bread');

assert(
   (Object.keys(groceryList.list).length == 2),
   "bread in the groceryList should equal to 5", 
   "6. "
);


// 5. Reflection

// What parts of your strategy worked? What problems did you face?
// --- I just had to reoganize my knowledge on javascript and also do some research.
// What questions did you have while coding? What resources did you find to help you answer them?
// --- Mostly some syntax quections since I was getting cinfused with Ruby suntax.  
// What concepts are you having trouble with, or did you just figure something out? If so, what?
// --- I just need to get used to writing and thinking in javascript whenever I need to.
// Did you learn any new skills or tricks?
// --- I knew about it, but for the first time I defined the whole object at once.
// How confident are you with each of the Learning Competencies?
// --- I feel much better but still not quite there yet.
// Which parts of the challenge did you enjoy?
// --- Triying new things and still passing the tests.
// Which parts of the challenge did you find tedious?
// Debugging.

File: week_8_and_9/4_Ruby/fibonacci_sequence/my_solution.rb
# U3.W8-9: 


# I worked on this challenge by myself.

# 2. Pseudocode

# => CONPUTE (5*num*num + 4) and set it equal to n1
# => SET n2 to n1 - 8
# => IF either or both of n1 and n2 are perfect square
# =>   THEN RETURN true
# =>   ELSE RETURN false


# 3. Initial Solution

def is_fibonacci?(num)
	n1 = (5*num*num + 4)
	n2 = n1 - 8
	is_perfect_square?(n1) || is_perfect_square?(n2)
end

def is_perfect_square?(num)
	sqrt = Math.sqrt(num)
    sqrt.is_a? Integer
end



# 4. Refactored Solution

def is_fibonacci?(num)
	last_two = [0, 1]
	while num > last_two.last
		x, y = last_two.pop(2)
		last_two.push(y, x+y)
	end
	return num == last_two.last ? true : false  
end

# 1. DRIVER TESTS GO BELOW THIS LINE

def assert 
	raise "Assertion Failed!" unless yield
	puts true
end

def random_fibonacci
   [0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584,4181,6765,10946].sample
end

assert { is_fibonacci?(random_fibonacci) == true }
assert { is_fibonacci?(8670007398507948658051921) == true }
assert { is_fibonacci?(random_fibonacci+100) == false }
assert { is_fibonacci?(927372692193078999171) == false }

# 5. Reflection 

# What parts of your strategy worked? What problems did you face?
# => First of all, I had to look up what exactly fibinacci sequence is and how to 
#    verify a number if it's a fibonacci number.
# What questions did you have while coding? What resources did you find to help you answer them?
# => How to determine if a number is a perfect suare. I got stuck that so I started looking 
#    online for different ideas.
# What concepts are you having trouble with, or did you just figure something out? If so, what?
# => I think I made the concept of fibonacci sequence too complicated in my head and
#    ended up overwhelming myself.
# Did you learn any new skills or tricks?
# => Since I could not figure out how to determine a perfect square number, I ended up
#    finding a different approach online. Although I modified it a bit, this refactored
#    was highly inspired by somebody else's work. But it's simple and easy to understand.
# How confident are you with each of the Learning Competencies?
# => I feel ok but not confident since I couldn't solve the challenge entirely on my own.
# Which parts of the challenge did you enjoy?
# => Learning different ways to approach a challenge.
# Which parts of the challenge did you find tedious?
# => Trying to deal with perfect square was intence!